### PR TITLE
[SPARK-24916][SQL] Fix type coercion for IN expression with subquery

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

The below SQL will throw `AnalysisException`. but it can success on Spark 2.1.x. This pr fix this issue.
```sql
CREATE TEMPORARY VIEW t4 AS SELECT * FROM VALUES
  (CAST(1 AS DOUBLE), CAST(2 AS STRING), CAST(3 AS STRING))
AS t1(t4a, t4b, t4c);

CREATE TEMPORARY VIEW t5 AS SELECT * FROM VALUES
  (CAST(1 AS DECIMAL(18, 0)), CAST(2 AS STRING), CAST(3 AS BIGINT))
AS t1(t5a, t5b, t5c);

SELECT * FROM t4
WHERE
(t4a, t4b, t4c) IN (SELECT t5a, t5b, t5c FROM t5);
```

I tested the different combinations of getting `commonTypes`, it looks like `findCommonTypeForBinaryComparison + findWiderTypeWithoutStringPromotionForTwo` is the best way. 

No. | commonTypes | desc
-- | -- | --
1 | findCommonTypeForBinaryComparison + findTightestCommonType | Can't compare `double` with `decimal`
2 | findCommonTypeForBinaryComparison + findWiderTypeForTwo | Seems good
3 | findWiderTypeForTwo | Can't compare `string` with `double`
4 | findCommonTypeForBinaryComparison + findWiderTypeWithoutStringPromotionForTwo | Same to No.2
5 | findWiderTypeWithoutStringPromotionForTwo | Can't compare `string` with `int`
6 | findCommonTypeForBinaryComparison + findTightestCommonType + findWiderTypeForDecimal | Same to No.2

The details can be found in the commit log.

## How was this patch tested?

unit tests

### CODE REVIEW

**Comment by mgaido91** [comment]
I think this is basically the same of what I proposed in https://github.com/apache/spark/pull/19635. Unfortunately, that PR got a bit stuck...

File: sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/inConversion.sql [hunk 1]
```diff
@@ -328,3 +328,159 @@ SELECT cast('2017-12-12 09:30:00' as date) in (cast('2017-12-12 09:30:00' as dat
 SELECT cast('2017-12-12 09:30:00' as date) in (cast('2017-12-12 09:30:00' as date), cast(1 as boolean)) FROM t;
 SELECT cast('2017-12-12 09:30:00' as date) in (cast('2017-12-12 09:30:00' as date), cast('2017-12-11 09:30:00.0' as timestamp)) FROM t;
 SELECT cast('2017-12-12 09:30:00' as date) in (cast('2017-12-12 09:30:00' as date), cast('2017-12-11 09:30:00' as date)) FROM t;
+
+SELECT * FROM t WHERE (cast(1 as tinyint)) IN (SELECT cast(1 as tinyint) FROM t);
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/typeCoercion/native/inConversion.sql]: Do we really need to test all the combinations? We need most of such logics should be tested in `findWiderTypeWithoutStringPromotionForTwo` and we could have just few end to end tests.

**Comment by wangyum** [comment]
Oh. It turns out that @dilipbiswal is talking about that PR.  I didn't find it in your recent PR. Let’s wait if the test can pass.

**Comment by wangyum** [comment]
workaround:
```sql
SELECT * FROM t4
WHERE
(t4a, t4b, t4c) IN (SELECT t5a, t5b, t5c FROM t5);

->

SELECT * FROM t4
WHERE
(t4a, t4b, t4c) IN (SELECT CAST(t5a as DOUBLE), CAST(t5b AS STRING), CAST(t5c AS STRING) FROM t5);
```

**Pull request closed without merging**